Example of TPC-H q1, q6 queries over Parquet files

majetideepak opened the pull request:
Add a framework to verify manually written TPC-H plans in Velox against the DuckDB TPC-H extension. Add TPC-H queries 1 and 6.

- Initialize DuckDB instance and the TPC-H data with scale factor `0.01` in the SetUpTestCase method.
- In the test, first write the `lineitem` table data into Parquet file. This requires `BUILD_PARQUET_EXTENSION`. Since Velox does not support `decimal` type yet, the corresponding columns are converted to `double`.
- Build Velox plan for a given query.
- Retrieve the corresponding DuckDB query using `TPCHExtension:::GetQuery()` API.
- Round double/decimal columns output with 2 scale for equality comparision.
- Use assertQuery() API to compare the results from Velox against the results from DuckDB.

Also fixes a non-deterministic bug with error `Reason: Unknown stream with ID 12`

mbasmanova commented on velox/dwio/parquet/tests/ParquetTpchTest.cpp:
```diff
@@ -0,0 +1,221 @@
+                              {0, 1},
+                              {"sum(quantity)",
+                               "sum(extendedprice)",
+                               "sum(p4)",
```
nit: you as use "AS" to specify an alias, e.g. "extendedprice * (1 - discount) as sum_disc_price", then "sum(sum_disc_price)"

mbasmanova commented on velox/dwio/parquet/tests/ParquetTpchTest.cpp:
```diff
@@ -0,0 +1,221 @@
+                          .planNode();
+
+  auto plan = PlanBuilder(1)
+                  .localPartition({0, 1}, {stage1})
```
You can use gather exchange here because we don't support running final aggregation multi-threaded, e.g. ".localPartition({}, {stage1})"

mbasmanova commented on velox/dwio/parquet/tests/ParquetTpchTest.cpp:
```diff
@@ -0,0 +1,221 @@
+  EXPECT_EQ(1, result.second.size());
+  auto output = materialize(result.second[0]);
+  ASSERT_EQ(10, output[0].size());
+  ASSERT_EQ(std::string("N"), std::string(output[0][0].value<StringView>()));
```
These assertions are pretty hard to read. Consider using assertQuery to compare the results with DuckDB. You can use TPC-H extension provided with DuckDB.

majetideepak replied:
The DuckDB TPC-H extension has only a single API to get results
https://github.com/duckdb/duckdb/blob/01bdd7ce8930a960ad4e9ded58f29efa8fb65cda/extension/tpch/include/tpch-extension.hpp#L23
```
//! Returns the CSV answer of a TPC-H query
static std::string GetAnswer(double sf, int query);
```
Instead of parsing CSV, the current change directly queries the Parquet files using the Parquet extension.
This does mean an additional step to copy the `parquet-extension.hpp` file similar to TPC-H extension.
Any thoughts?

mbasmanova replied:
@Mytherin @hannesmuehleisen What would be a proper way to run a TPC-H query in DuckDB and get results as a vector, not CSV?

mbasmanova replied:
@majetideepak Have you tried fetching query test using "static std::string GetQuery(int query);" API and running it the same way we run other queries, e.g. via assertQuery?

majetideepak replied:
@mbasmanova `::duckdb::TPCHExtension::GetQuery(1)` returns the following string. 
```
"SELECT\n    l_returnflag,\n    l_linestatus,\n    sum(l_quantity) AS sum_qty,\n    sum(l_extendedprice) AS sum_base_price,\n    sum(l_extendedprice * (1 - l_discount)) AS sum_disc_price,\n    sum(l_extendedprice * (1 - l_discount) * (1 + l_tax)) AS sum_charge,\n    avg(l_quantity) AS avg_qty,\n    avg(l_extendedprice) AS avg_price,\n    avg(l_discount) AS avg_disc,\n    count(*) AS count_order\nFROM\n    lineitem\nWHERE\n    l_shipdate <= CAST('1998-09-02' AS date)\nGROUP BY\n    l_returnflag,\n    l_linestatus\nORDER BY\n    l_returnflag,\n    l_linestatus;\n"
```
We can use this only after creating a duckdb table `ineitem` first correct? That means we will have to create the tpc-h `lineitem` data manually and pass it to `DuckDbQueryRunner::createTable(lineitem, data)`. This is a redundant step correct right?

Did I miss something?

majetideepak replied:
Alternatively, we could add `DuckDbQueryRunner::createTableFromParquet(lineitem, filePaths)` and implement it as
```
CREATE TABLE lineitem AS SELECT * FROM parquet_scan('');
```
But we still need to add the Parquet extension for this to work.

majetideepak replied:
The other issue with the current `assertQuery` API is that there is no way to specify `params.maxDrivers = 4;` and the source plan id must always be `0`

majetideepak replied:
I figured out how to load the TPCH data into DuckDB. We can invoke `::duckdb::TPCHExtension::GetQuery(1)` and should be good.
```
    void initTPCH() {
      db_.LoadExtension<::duckdb::TPCHExtension>();
      ::duckdb::Connection con(db_);
      con.Query("CALL dbgen(sf=0.01)");
    }
```

mbasmanova replied:
That's great.

mbasmanova replied:
> The other issue with the current `assertQuery` API is that there is no way to specify `params.maxDrivers = 4;` and the source plan id must always be `0`

There is a version of assertQuery that takes CursorParameters and addSplits lambda in velox/exec/tests/utils/QueryAssertions.h. This allows you to specify params.maxDrivers and use whatever plan node ID you want when adding splits.

```
std::shared_ptr<Task> assertQuery(
    const CursorParameters& params,
    std::function<void(exec::Task*)> addSplits,
    const std::string& duckDbSql,
    DuckDbQueryRunner& duckDbQueryRunner,
    std::optional<std::vector<uint32_t>> sortingKeys = std::nullopt);
```

mbasmanova replied:
Ideally, we would use DuckDB's Parquet writer to generate Parquet files on the fly instead of checking them in.

majetideepak replied:
Makes sense. I will try generating the files.

majetideepak commented:
@mbasmanova Can you take another look? Thanks.

majetideepak commented on velox/dwio/parquet/tests/ParquetTpchTest.cpp:
```diff
@@ -0,0 +1,304 @@
+  // location.
+  std::string writeDuckDBTPCHLineitemTableToParquet() {
+    constexpr std::string_view tableName("lineitem");
+    // Lineitem SF=0.01 has about 64K rows
```
`60175` rows precisely.

majetideepak commented on velox/dwio/parquet/reader/duckdb/InputStreamFileSystem.h:
```diff
@@ -179,7 +179,7 @@ class InputStreamFileSystem : public ::duckdb::FileSystem {
 
  private:
-  int nextStreamId_ = 1;
+  std::atomic<int> nextStreamId_ = 1;
```
While working on this PR, I occasionally noticed the following core. I was able to reproduce the core in a 500 run test. With this fix, I do not see this issue anymore.
CI run that also hit this core
https://circleci.com/api/v1.1/project/github/facebookincubator/velox/22031/output/109/0?file=true&allocation-id=61f4846f768d232005f83141-0-build%2F3073EF21
```
WARNING: Logging before InitGoogleLogging() is written to STDERR
E0129 00:26:54.578812  9820 Exceptions.h:68] ../.././velox/dwio/parquet/reader/duckdb/InputStreamFileSystem.h46OpenFileit != lockedStreams->end()Unknown stream with ID 12RUNTIMEINVALID_STATE
E0129 00:26:54.579144  9820 Exceptions.h:68] ../.././velox/dwio/parquet/reader/duckdb/InputStreamFileSystem.h63CloseStreamit != lockedStreams->end()Unknown stream with ID 12RUNTIMEINVALID_STATE
terminate called after throwing an instance of 'facebook::velox::VeloxRuntimeError'
unknown file: Failure
C++ exception with description "Exception: VeloxRuntimeError
Error Source: RUNTIME
Error Code: INVALID_STATE
Reason: Unknown stream with ID 12
Retriable: False
Expression: it != lockedStreams->end()
Function: OpenFile
File: ../.././velox/dwio/parquet/reader/duckdb/InputStreamFileSystem.h
Line: 46
Stack trace:
```

mbasmanova replied:
@majetideepak Thank you for debugging and fixing. It would be nice to mention this fix in the PR description.

mbasmanova commented on velox/exec/tests/utils/QueryAssertions.cpp:
```diff
@@ -425,6 +425,23 @@ void DuckDbQueryRunner::createTable(
   }
 }
 
+void DuckDbQueryRunner::initTPCH(double sf) {
```
naming: initializeTpch(double scaleFactor)

mbasmanova commented on velox/dwio/parquet/tests/ParquetTpchTest.cpp:
```diff
@@ -0,0 +1,306 @@
+    return filePath;
+  }
+
+  static std::shared_ptr<DuckDbQueryRunner> duckDb_;
```
Why can't we re-use  duckDbQueryRunner_ member variable from the base class?

majetideepak replied:
The idea is to load the TPC-H data into DuckDB only once for the entire test suite here. That is why I declared it as static and set it up inside `static void SetupTestCase`.
The base class will instantiate `DuckDbQueryRunner` for each test.

majetideepak replied:
To provide more context, I wanted to invoke `initializeTpch(kTpchScaleFactor);` only once per suite as it takes about `4 seconds` to load the data. If this is not a deal, I can use re-use the member variable.

mbasmanova replied:
Thank you for explaining. It is a bit confusing to have duckDb_ and duckDbQueryRunner_ in the base class. It also means that assertQuery methods from the base class cannot be used easily. Would it not work to simply call initializeTpch() on duckDbQueryRunner_ once?

mbasmanova commented on velox/dwio/parquet/tests/ParquetTpchTest.cpp:
```diff
@@ -0,0 +1,306 @@
+      constexpr double tpchsf = 0.01;
+      duckDb_->initTPCH(tpchsf);
+    }
+    functions::prestosql::registerAllFunctions();
```
Should we call OperatorTestBase::SetUpTestCase instead?

mbasmanova commented on velox/dwio/parquet/tests/ParquetTpchTest.cpp:
```diff
@@ -0,0 +1,306 @@
+    return date.days();
+  }
+
+  // For a given file at filePath, add splits equal to numSplits.
```
"add splits equal to numSplits" -> "create 'numSplits' number of splits" or "Split file at a given path into 'numSplits' splits."

mbasmanova commented on velox/dwio/parquet/tests/ParquetTpchTest.cpp:
```diff
@@ -0,0 +1,306 @@
+  std::vector<std::shared_ptr<connector::hive::HiveConnectorSplit>> makeSplits(
+      const std::string& filePath,
+      int64_t numSplits) const {
+    LocalReadFile lfs(filePath);
```
nit: Any particular reason not to use fs::file_size ? 

```
#if __has_include("filesystem")
#include <filesystem>
namespace fs = std::filesystem;
#else
#include <experimental/filesystem>
namespace fs = std::experimental::filesystem;
#endif
```

mbasmanova commented on velox/dwio/parquet/tests/ParquetTpchTest.cpp:
```diff
@@ -0,0 +1,306 @@
+    const int splitSize = fileSize / numSplits;
+    std::vector<std::shared_ptr<connector::hive::HiveConnectorSplit>> splits;
+
+    // Add all the splits except the last one.
```
I think you can just add all splits in a loop. No need to treat last split in any special way.

mbasmanova commented on velox/dwio/parquet/tests/ParquetTpchTest.cpp:
```diff
@@ -0,0 +1,306 @@
+
+  // Write the Lineitem TPCH table to a Parquet file and return the file
+  // location.
+  std::string writeDuckDBTPCHLineitemTableToParquet() const {
```
This method name is very long and hard to read. Perhaps, add tableName argument to this method to remove "Lineitem" from the method name.

```
std::string writeTpchTableAsParquet(const std::string& tableName)
```

mbasmanova commented on velox/dwio/parquet/tests/ParquetTpchTest.cpp:
```diff
@@ -0,0 +1,306 @@
+std::shared_ptr<DuckDbQueryRunner> ParquetTpchTest::duckDb_ = nullptr;
+
+TEST_F(ParquetTpchTest, tpchQ1) {
+  const auto& lineitemFilePath = writeDuckDBTPCHLineitemTableToParquet();
```
nit: filePath for brevity

mbasmanova commented on velox/dwio/parquet/tests/ParquetTpchTest.cpp:
```diff
@@ -0,0 +1,306 @@
+TEST_F(ParquetTpchTest, tpchQ1) {
+  const auto& lineitemFilePath = writeDuckDBTPCHLineitemTableToParquet();
+
+  // TPCH Q1
```
this comment is redundant

mbasmanova commented on velox/dwio/parquet/tests/ParquetTpchTest.cpp:
```diff
@@ -0,0 +1,306 @@
+  const auto& lineitemFilePath = writeDuckDBTPCHLineitemTableToParquet();
+
+  // TPCH Q1
+  auto rowType = ROW(
```
perhaps, add a member variable with the lineitem schema, then add a helper method to get types for selected columns, e.g.

```
auto rowType = getLineitemColumns({"returnflag",
       "linestatus",
       "quantity",
       "extendedprice",
       "discount",
       "tax",
       "shipdate"});
```

mbasmanova commented on velox/dwio/parquet/tests/ParquetTpchTest.cpp:
```diff
@@ -0,0 +1,306 @@
+  params.planNode = std::move(plan);
+  bool noMoreSplits = false;
+  auto duckDbSql = duckDb_->GetQuery(1);
+  // Additional step for double type result verification.
```
perhaps, fold this logic into the helper method duckDb_->GetQuery

mbasmanova reviewed (commented):
@majetideepak Thank you for putting this together. Some comments below.

majetideepak commented on velox/exec/tests/utils/QueryAssertions.cpp:
```diff
@@ -581,22 +598,7 @@ std::shared_ptr<Task> assertQuery(
-        duckDbQueryRunner);
-  }
-  return result.first->task();
+  return assertQuery(params, addSplits, duckDbSql, duckDbQueryRunner);
```
I missed the last argument. Thanks for catching this. But the code here is the same as

```
std::shared_ptr<Task> assertQuery(
    const CursorParameters& params,
    std::function<void(exec::Task*)> addSplits,
    const std::string& duckDbSql,
    DuckDbQueryRunner& duckDbQueryRunner,
    std::optional<std::vector<uint32_t>> sortingKeys) {
  auto [cursor, actualResults] = readCursor(params, addSplits);

  if (sortingKeys) {
    assertResultsOrdered(
        actualResults,
        params.planNode->outputType(),
        duckDbSql,
        duckDbQueryRunner,
        sortingKeys.value());
  } else {
    assertResults(
        actualResults,
        params.planNode->outputType(),
        duckDbSql,
        duckDbQueryRunner);
  }
  return cursor->task();
}
```

majetideepak reviewed (commented):
@mbasmanova thanks for the feedback!

mbasmanova commented on velox/exec/tests/utils/QueryAssertions.h:
```diff
@@ -59,6 +63,14 @@ class DuckDbQueryRunner {
     return allRows;
   }
 
+  // returns the DuckDB TPC-H Extension Query as string for a given 'queryNo'
```
Comments should start with a capital letter and end with a period. E.g. "Returns the ..."

mbasmanova commented on velox/exec/tests/utils/QueryAssertions.cpp:
```diff
@@ -425,21 +426,34 @@ void DuckDbQueryRunner::createTable(
+  execute(query);
+}
+
+void verifyDuckDBResult(DuckDBQueryResult& result, std::string_view sql) {
```
- result can be passed as a const reference, e.g. const DuckDBQueryResult& result
- put this function into an anonymous namespace

mbasmanova commented on velox/dwio/parquet/tests/ParquetTpchTest.cpp:
```diff
@@ -0,0 +1,352 @@
+  }
+
+  RowTypePtr getLineitemColumns(
+      std::vector<std::string>&& outputColumnNames) const {
```
I'd drop && since there is no need to force the caller to std::move the names. I'd also shorten the parameter name to "names" for readability.

```
RowTypePtr getLineitemColumns(std::vector<std::string> names) const
```

mbasmanova commented on velox/dwio/parquet/tests/ParquetTpchTest.cpp:
```diff
@@ -0,0 +1,352 @@
+    return exec::test::assertQuery(
+        params,
+        [&](exec::Task* task) {
+          addSplitsToTask(task, noMoreSplits, {filePath}, sourcePlanNodeId);
```
addSplitsToTask is used only in one place; it might be cleaner to inline that logic

mbasmanova commented on velox/dwio/parquet/tests/ParquetTpchTest.cpp:
```diff
@@ -0,0 +1,352 @@
+
+// output of getTpchQuery() has a new line and semi-colon that need to be
+// removed in order to use the query string in a subquery
+inline void prepareQueryStringForSubquery(std::string& queryString) {
```
Would it make sense to fold this into getTpchQuery() method?

mbasmanova reviewed (commented):
@majetideepak Looks good. A couple of remaining questions.

majetideepak commented:
@mbasmanova I address all the review comments except for the `static duckDb_` vs the `member duckDbQueryRunner_`.
My understanding is that the member variables get instantiated for each test case. So here both Q1 and Q6 would get a different instance of `duckDbQueryRunner_`. Invoking `initializeTpch()` in one will not persist for another.
The only way to persist across tests is to create a static instance like here.
I am also not pleased with having both. How about making `ParquetTpchTest` be derived directly from `public testing::Test` and not `HiveConnectorTestBase`?
Currently, this test is also using AssertQuery directly from `queryAssertions.h` and not the one in HiveConnectorTestBase as it lacks API to take a lambda for adding splits. So there is little overlap.
However, this does come with the overhead of duplicating `HiveConnectorTestBase::SetUp()`

mbasmanova commented:
> How about making ParquetTpchTest be derived directly from public testing::Test and not HiveConnectorTestBase?

This would work. An alternative might be to allow classes derived from OperatorTestBase to provide their own instance of DuckDbQueryRunner.

majetideepak commented:
@mbasmanova I went with using  `public testing::Test` as the base class. There was not much to gain from using `OperatorTestBase` as the base class.

mbasmanova reviewed (approved):
@majetideepak Thanks.

mbasmanova commented:
@majetideepak Deepak, I'm getting an error in Phabricator. Will need to look into it.

```
stderr: In file included from velox/exec/tests/utils/OperatorTestBase.cpp:17:
In file included from velox/exec/tests/utils/OperatorTestBase.h:22:
In file included from velox/exec/tests/utils/QueryAssertions.h:21:
velox/external/duckdb/tpch/include/tpch-extension.hpp:11:10: fatal error: 'duckdb.hpp' file not found
#include "duckdb.hpp"
         ^~~~~~~~~~~~
1 error generated.
```

majetideepak commented:
@mbasmanova I refactored a bit to remove `velox/external/duckdb/tpch/include/tpch-extension.hpp` from the header file. I also fixed the path inside `tpch-extension.hpp`. This might help the build. Currently, `tpch-extension.hpp` header is being used inside `velox/duckdb/conversion/DuckWrapper.cpp`
